File: server/src/classes/RtEngine/handlers/sendToTopic.js
export default async function (topic, event, data) {
	if (!this.engine) {
		throw new Error("Engine not initialized")
	}

	if (
		this.server.nats &&
		typeof this.server.nats.operations?.sendToTopic === "function"
	) {
		return await this.server.nats.operations.sendToTopic(topic, event, data)
	}

	// publish message to topic with structured payload
	return this.engine.app.publish(
		topic,
		this.encode({
			topic: topic,
			event: event,
			data: data,
		}),
	)
}

File: gateway/core/websocket/nats_operations/send_to_topic.go
package nats_operations

import (
	"log"
	"ultragateway/structs"

	"github.com/bytedance/sonic"
	"github.com/lxzan/gws"
	"github.com/nats-io/nats.go"
)

func (context *Instance) SendToTopic(conn *gws.Conn, connCtx *structs.WSConnectionCtx, msg *nats.Msg) *structs.OperationResult {
	dataNode, err := sonic.Get(msg.Data, "data")

	if err != nil {
		return &structs.OperationResult{
			Ok:    false,
			Error: "Failed to parse payload",
		}
	}

	targetTopic, err := dataNode.Get("topic").String()

	if err != nil || targetTopic == "" {
		return &structs.OperationResult{
			Ok:    false,
			Error: "Invalid payload or missing topic",
		}
	}

	data, err := dataNode.MarshalJSON()

	if err != nil {
		return &structs.OperationResult{
			Ok:    false,
			Error: "Failed to parse payload [data]",
		}
	}

	broadcaster := gws.NewBroadcaster(gws.OpcodeText, data)
	defer broadcaster.Close()

	if IsDebug {
		log.Printf("Sending a event to topic[%s]", targetTopic)
	}

	context.PubSub.Publish(targetTopic, broadcaster)

	return &structs.OperationResult{Ok: true}
}
